File: wave_lang/kernel/wave/asm/metadata_emitter.py
# Copyright 2025 The IREE Authors
#
# Licensed under the Apache License, Version 2.0 with LLVM Exceptions.
# See https://llvm.org/LICENSE.txt for license information.
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
"""
Metadata Emitter for AMDGCN kernels.

This module handles the emission of assembler directives for kernel
prologue (before instructions) and epilogue (after instructions).

These are NOT instructions - they are assembler directives like:
- .amdgcn_target
- .amdhsa_kernel / .end_amdhsa_kernel
- .amdgpu_metadata

The kernel body instructions are handled by KernelGenerator.

Architecture:
    ┌─────────────────────────────────────────────────────────────┐
    │  MetadataEmitter.emit_prologue()                            │
    │  - .amdgcn_target, .amdhsa_kernel, etc.                     │
    └─────────────────────────────────────────────────────────────┘
                                 │
                                 ▼
    ┌─────────────────────────────────────────────────────────────┐
    │  KernelGenerator.generate()                                 │
    │  - All instructions (including kernargs, s_endpgm)          │
    └─────────────────────────────────────────────────────────────┘
                                 │
                                 ▼
    ┌─────────────────────────────────────────────────────────────┐
    │  MetadataEmitter.emit_epilogue()                            │
    │  - .amdgpu_metadata                                         │
    └─────────────────────────────────────────────────────────────┘
"""

from dataclasses import dataclass
from typing import List, Tuple

from .utils import normalize_wg_size
from .abi import system_vgpr_workitem_id_from_wg_size


def get_register_granularity(targetid: str) -> Tuple[int, int]:
    """Get VGPR and SGPR allocation granularity for target."""
    # Canonical rule for current targets:
    # - gfx94x/gfx95x: VGPRs allocated in groups of 8, SGPRs in groups of 8
    # - older GFX9: VGPRs in groups of 4, SGPRs in groups of 8
    if any(t in targetid for t in ["gfx940", "gfx942", "gfx950"]):
        return 8, 8
    return 4, 8


@dataclass
class KernelMetadata:
    """Metadata about a kernel for prologue/epilogue generation."""

    name: str
    targetid: str
    codeobj: str = "5"

    # Workgroup configuration
    wg_size: Tuple[int, int, int] = (64, 1, 1)
    subgroup_size: int = 64

    # Workgroup ID requirements
    needs_wgid_x: bool = False
    needs_wgid_y: bool = False
    needs_wgid_z: bool = False

    # Resource usage (filled in after compilation)
    vgprs_used: int = 0
    sgprs_used: int = 0
    agprs_used: int = 0
    lds_size_bytes: int = 0

    # Kernel arguments
    num_args: int = 0

    # SRD upper word constant
    SRD127_96: int = 0x20000


class MetadataEmitter:
    """
    Emits assembler directives for kernel prologue and epilogue.

    This class handles only metadata/directives, not instructions.
    Instructions (including kernargs loading, s_endpgm) are handled
    by KernelGenerator through the kernel IR.
    """

    def __init__(self, metadata: KernelMetadata):
        self.metadata = metadata
        self._lines: List[str] = []

    def emit_prologue(self) -> List[str]:
        """
        Emit kernel prologue with metadata directives.

        Returns list of assembly lines for the prologue.
        """
        m = self.metadata
        lines = []

        # Target specification
        lines.append(f'.amdgcn_target "amdgcn-amd-amdhsa--{m.targetid}"')
        lines.append(".text")
        lines.append(f".protected {m.name}")
        lines.append(f".globl {m.name}")
        lines.append(".p2align 8")
        lines.append(f".type {m.name},@function\n")

        # Kernel descriptor section
        lines.append(".section .rodata,#alloc")
        lines.append(".p2align 6")
        lines.append(f".amdhsa_kernel {m.name}")
        lines.append("  .amdhsa_user_sgpr_kernarg_segment_ptr 1")

        # User SGPR count
        lines.append("  .amdhsa_user_sgpr_count 2")

        # Resource usage (placeholders to be patched later)
        lines.append("  .amdhsa_accum_offset 0")  # patched later
        lines.append("  .amdhsa_next_free_vgpr 0")  # patched later
        lines.append("  .amdhsa_next_free_sgpr 0")  # patched later
        lines.append("  .amdhsa_group_segment_fixed_size 0")  # patched later
        lines.append("  .amdhsa_private_segment_fixed_size 0")

        # Workgroup ID requests
        lines.append(
            f"  .amdhsa_system_sgpr_workgroup_id_x {1 if m.needs_wgid_x else 0}"
        )
        lines.append(
            f"  .amdhsa_system_sgpr_workgroup_id_y {1 if m.needs_wgid_y else 0}"
        )
        lines.append(
            f"  .amdhsa_system_sgpr_workgroup_id_z {1 if m.needs_wgid_z else 0}"
        )

        # Workitem ID VGPRs
        system_vgpr_workitem_id = self._get_system_vgpr_workitem_id()
        lines.append(f"  .amdhsa_system_vgpr_workitem_id {system_vgpr_workitem_id}")

        # Float mode
        lines.append("  .amdhsa_float_denorm_mode_32 3")
        lines.append("  .amdhsa_float_denorm_mode_16_64 3")

        lines.append(".end_amdhsa_kernel")
        lines.append(".text\n")

        # SRD constant
        lines.append("# SRD upper word (gfx9xx): data_format=4 => 0x20000")
        lines.append(f".set Srd127_96, {m.SRD127_96}\n")

        # Kernel label
        lines.append(f"{m.name}:")

        return lines

    def emit_epilogue(self) -> List[str]:
        """
        Emit kernel epilogue with YAML metadata.

        Returns list of assembly lines for the epilogue.
        """
        m = self.metadata
        lines = []

        # Empty line after s_endpgm
        lines.append("")

        # Build YAML args
        args_yaml = []
        for i in range(m.num_args):
            args_yaml.append(
                f"""      - .name: arg{i}_ptr
        .size: 8
        .offset: {i*8}
        .value_kind: global_buffer
        .value_type: i8*"""
            )
        args_yaml_string = "\n".join(args_yaml)

        amdhsa_minor = "2" if m.codeobj == "5" else "1"
        wg_x, wg_y, wg_z = normalize_wg_size(m.wg_size)

        metadata = f"""
.amdgpu_metadata
---
amdhsa.version:
  - 1
  - {amdhsa_minor}
amdhsa.kernels:
  - .name: {m.name}
    .symbol: '{m.name}.kd'
    .language: OpenCL C
    .language_version:
      - 1
      - 2
    .kernarg_segment_size: {m.num_args * 8}
    .group_segment_fixed_size: {m.lds_size_bytes}
    .private_segment_fixed_size: 0
    .kernarg_segment_align: 8
    .wavefront_size: {m.subgroup_size}
    .sgpr_count: {m.sgprs_used}
    .vgpr_count: {m.vgprs_used}
    .max_flat_workgroup_size: {wg_x * wg_y * wg_z}
    .args:
{args_yaml_string}
...
.end_amdgpu_metadata"""

        lines.append(metadata)
        return lines

    def _emit_workgroup_id_sgprs(self, lines: List[str]) -> None:
        """Emit workgroup ID SGPR configuration."""
        m = self.metadata
        if m.needs_wgid_x:
            lines.append("  .amdhsa_system_sgpr_workgroup_id_x 1")
        if m.needs_wgid_y:
            lines.append("  .amdhsa_system_sgpr_workgroup_id_y 1")
        if m.needs_wgid_z:
            lines.append("  .amdhsa_system_sgpr_workgroup_id_z 1")

    def _get_system_vgpr_workitem_id(self) -> int:
        """Determine system VGPR workitem ID mode based on workgroup size."""
        return system_vgpr_workitem_id_from_wg_size(self.metadata.wg_size)

    @staticmethod
    def patch_resource_usage(
        lines: List[str],
        vgprs_used: int,
        sgprs_used: int,
        agprs_used: int,
        lds_size_bytes: int,
        targetid: str,
    ) -> List[str]:
        """
        Patch resource usage in the prologue lines.

        This is called after compilation to fill in the actual values.
        """
        vgpr_granularity, sgpr_granularity = get_register_granularity(targetid)

        # Round up to allocation granularity
        vgprs_used = (
            (vgprs_used + vgpr_granularity - 1) // vgpr_granularity
        ) * vgpr_granularity
        sgprs_used = (
            (sgprs_used + sgpr_granularity - 1) // sgpr_granularity
        ) * sgpr_granularity

        # Compute accum_offset
        accum_offset = max(4, vgprs_used)

        # Handle AGPRs
        if agprs_used > 0:
            agprs_used = (
                (agprs_used + vgpr_granularity - 1) // vgpr_granularity
            ) * vgpr_granularity
            total_arch_vgprs = accum_offset + agprs_used
            vgprs_used = max(vgprs_used, total_arch_vgprs)

        # Patch the lines
        txt = "\n".join(lines)
        txt = txt.replace(
            "  .amdhsa_next_free_vgpr 0", f"  .amdhsa_next_free_vgpr {vgprs_used}"
        )
        txt = txt.replace(
            "  .amdhsa_accum_offset 0", f"  .amdhsa_accum_offset {accum_offset}"
        )
        txt = txt.replace(
            "  .amdhsa_next_free_sgpr 0", f"  .amdhsa_next_free_sgpr {sgprs_used}"
        )
        txt = txt.replace(
            "  .amdhsa_group_segment_fixed_size 0",
            f"  .amdhsa_group_segment_fixed_size {lds_size_bytes}",
        )

        return txt.splitlines()


def create_metadata(
    name: str,
    targetid: str,
    codeobj: str = "5",
    wg_size: Tuple[int, int, int] = (64, 1, 1),
    subgroup_size: int = 64,
    needs_wgid: Tuple[bool, bool, bool] = (False, False, False),
    num_args: int = 0,
) -> KernelMetadata:
    """Create kernel metadata from parameters."""
    return KernelMetadata(
        name=name,
        targetid=targetid,
        codeobj=codeobj,
        wg_size=normalize_wg_size(wg_size),
        subgroup_size=subgroup_size,
        needs_wgid_x=needs_wgid[0],
        needs_wgid_y=needs_wgid[1],
        needs_wgid_z=needs_wgid[2],
        num_args=num_args,
    )

File: waveasm/lib/Transforms/MetadataEmitter.cpp
// Copyright 2025 The Wave Authors
//
// Licensed under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

//===----------------------------------------------------------------------===//
// Instruction Formatter and Metadata Emitter
//
// InstructionFormatter: formats individual AMDGCN assembly instructions.
// MetadataEmitter: emits kernel prologue (.text, symbol), epilogue (kernel
//   descriptor, YAML metadata) for AMDGCN assembly output.
//===----------------------------------------------------------------------===//

#include "waveasm/Dialect/WaveASMOps.h"
#include "waveasm/Dialect/WaveASMTypes.h"
#include "waveasm/Transforms/AssemblyEmitter.h"

using namespace mlir;

namespace waveasm {

//===----------------------------------------------------------------------===//
// Instruction Formatter Implementation
//===----------------------------------------------------------------------===//

std::string InstructionFormatter::format(llvm::StringRef name,
                                         llvm::ArrayRef<std::string> operands) {
  std::string result = "  " + name.str();

  if (!operands.empty()) {
    result += " ";
    for (size_t i = 0; i < operands.size(); ++i) {
      if (i > 0)
        result += ", ";
      result += operands[i];
    }
  }

  return result;
}

std::string InstructionFormatter::formatLabel(llvm::StringRef label) {
  return label.str() + ":";
}

std::string InstructionFormatter::formatComment(llvm::StringRef text) {
  return "  ; " + text.str();
}

std::string InstructionFormatter::formatWaitcnt(std::optional<int64_t> vmcnt,
                                                std::optional<int64_t> lgkmcnt,
                                                std::optional<int64_t> expcnt) {
  std::string result = "  s_waitcnt";
  llvm::SmallVector<std::string> counts;

  if (vmcnt.has_value()) {
    counts.push_back("vmcnt(" + std::to_string(*vmcnt) + ")");
  }
  if (lgkmcnt.has_value()) {
    counts.push_back("lgkmcnt(" + std::to_string(*lgkmcnt) + ")");
  }
  if (expcnt.has_value()) {
    counts.push_back("expcnt(" + std::to_string(*expcnt) + ")");
  }

  if (!counts.empty()) {
    result += " ";
    for (size_t i = 0; i < counts.size(); ++i) {
      if (i > 0)
        result += " & ";
      result += counts[i];
    }
  }

  return result;
}

std::string InstructionFormatter::formatBarrier() { return "  s_barrier"; }

std::string InstructionFormatter::formatEndpgm() { return "  s_endpgm"; }

std::string InstructionFormatter::formatRaw(llvm::StringRef text) {
  return "  " + text.str();
}

//===----------------------------------------------------------------------===//
// Metadata Emitter Implementation
//===----------------------------------------------------------------------===//

MetadataEmitter::MetadataEmitter(ProgramOp program, TargetAttrInterface target)
    : program(program), target(target) {}

llvm::SmallVector<std::string> MetadataEmitter::emitPrologue() {
  llvm::SmallVector<std::string> lines;

  lines.push_back(target.getTargetDirective().str());
  lines.push_back("");
  lines.push_back(".text");
  lines.push_back("");

  std::string symName = getKernelName(program).str();
  lines.push_back(".protected " + symName);
  lines.push_back(".globl " + symName);
  lines.push_back(".p2align 8");
  lines.push_back(".type " + symName + ",@function");
  lines.push_back(symName + ":");

  return lines;
}

llvm::SmallVector<std::string> MetadataEmitter::emitEpilogue(int64_t peakVGPRs,
                                                             int64_t peakSGPRs,
                                                             int64_t peakAGPRs,
                                                             int64_t ldsSize) {
  llvm::SmallVector<std::string> lines;

  auto descriptor =
      emitKernelDescriptor(peakVGPRs, peakSGPRs, peakAGPRs, ldsSize);
  lines.append(descriptor.begin(), descriptor.end());
  lines.push_back("");

  auto metadata = emitMetadataYAML(peakVGPRs, peakSGPRs, peakAGPRs, ldsSize);
  lines.append(metadata.begin(), metadata.end());

  return lines;
}

/// Scan program operations to detect system register usage
static void scanSystemRegisterUsage(ProgramOp program, bool &usesWorkgroupIdX,
                                    bool &usesWorkgroupIdY,
                                    bool &usesWorkgroupIdZ,
                                    bool &usesWorkitemId) {
  usesWorkgroupIdX = false;
  usesWorkgroupIdY = false;
  usesWorkgroupIdZ = false;
  usesWorkitemId = false;

  auto targetAttr = program.getTarget();
  auto targetKind = targetAttr.getTargetKind();
  bool isGfx950 = llvm::isa<GFX950TargetAttr>(targetKind);

  int64_t numArgs = 2;
  if (auto numArgsAttr =
          program->getAttrOfType<IntegerAttr>("num_kernel_args")) {
    numArgs = numArgsAttr.getInt();
  }

  int64_t userSgprCount = 2;
  if (isGfx950) {
    // Hardware limits user SGPRs to 16 on gfx950.
    userSgprCount = std::min(int64_t(16), 2 + numArgs * 2);
  }

  int64_t wgIdXIndex = userSgprCount;
  int64_t wgIdYIndex = userSgprCount + 1;
  int64_t wgIdZIndex = userSgprCount + 2;

  program.walk([&](Operation *op) {
    if (auto precolored = dyn_cast<PrecoloredSRegOp>(op)) {
      int64_t idx = precolored.getIndex();
      int64_t size = precolored.getSize();
      if (size == 1) {
        if (idx == wgIdXIndex)
          usesWorkgroupIdX = true;
        else if (idx == wgIdYIndex)
          usesWorkgroupIdY = true;
        else if (idx == wgIdZIndex)
          usesWorkgroupIdZ = true;
      }
    }
  });
}

llvm::SmallVector<std::string>
MetadataEmitter::emitKernelDescriptor(int64_t peakVGPRs, int64_t peakSGPRs,
                                      int64_t peakAGPRs, int64_t ldsSize) {
  llvm::SmallVector<std::string> lines;
  std::string symName = getKernelName(program).str();

  bool usesWorkgroupIdX, usesWorkgroupIdY, usesWorkgroupIdZ, usesWorkitemId;
  scanSystemRegisterUsage(program, usesWorkgroupIdX, usesWorkgroupIdY,
                          usesWorkgroupIdZ, usesWorkitemId);

  lines.push_back("");
  lines.push_back(".section .rodata,#alloc");
  lines.push_back(".p2align 6");
  lines.push_back(".amdhsa_kernel " + symName);

  int64_t ldsBlocks = (ldsSize + 127) / 128;
  lines.push_back("  .amdhsa_group_segment_fixed_size " +
                  std::to_string(ldsBlocks * 128));
  lines.push_back("  .amdhsa_private_segment_fixed_size 0");

  // Kernarg size must account for all args the runtime packs: buffer ptrs +
  // dynamic dims + stride args. Use 2x base to safely accommodate strides.
  {
    int64_t numArgsForSize = 2;
    if (auto numArgsAttr =
            program->getAttrOfType<IntegerAttr>("num_kernel_args")) {
      numArgsForSize = numArgsAttr.getInt();
    }
    int64_t kernargSize = numArgsForSize * 8 * 2;
    lines.push_back("  .amdhsa_kernarg_size " + std::to_string(kernargSize));
  }

  auto targetAttr = program.getTarget();
  auto targetKind = targetAttr.getTargetKind();
  int64_t preloadLength = program.getKernargPreloadLength();
  bool usePreloading = llvm::isa<GFX950TargetAttr>(targetKind);

  if (usePreloading && preloadLength == 0) {
    int64_t numArgs = 2;
    if (auto numArgsAttr =
            program->getAttrOfType<IntegerAttr>("num_kernel_args")) {
      numArgs = numArgsAttr.getInt();
    }
    // Hardware limits user SGPRs to 16 on gfx950 (2 for kernarg ptr + 14 max
    // preloaded). Overflow args are loaded via explicit s_load in the prologue.
    preloadLength = std::min(int64_t(14), numArgs * 2);
  }

  int64_t userSgprCount = 2;
  if (usePreloading && preloadLength > 0) {
    userSgprCount = 2 + preloadLength;
  }

  lines.push_back("  .amdhsa_user_sgpr_count " + std::to_string(userSgprCount));
  lines.push_back("  .amdhsa_user_sgpr_dispatch_ptr 0");
  lines.push_back("  .amdhsa_user_sgpr_queue_ptr 0");
  lines.push_back("  .amdhsa_user_sgpr_kernarg_segment_ptr 1");
  lines.push_back("  .amdhsa_user_sgpr_dispatch_id 0");

  if (usePreloading && preloadLength > 0) {
    lines.push_back("  .amdhsa_user_sgpr_kernarg_preload_length " +
                    std::to_string(preloadLength));
    lines.push_back("  .amdhsa_user_sgpr_kernarg_preload_offset 0");
  }

  lines.push_back("  .amdhsa_user_sgpr_private_segment_size 0");
  lines.push_back("  .amdhsa_uses_dynamic_stack 0");
  lines.push_back("  .amdhsa_enable_private_segment 0");

  int64_t vgprGranularity = 4;
  if (llvm::isa<GFX942TargetAttr, GFX950TargetAttr>(targetKind)) {
    vgprGranularity = 8;
  }
  int64_t nextFreeVGPR =
      ((peakVGPRs + vgprGranularity - 1) / vgprGranularity) * vgprGranularity;
  int64_t nextFreeAGPR =
      ((peakAGPRs + vgprGranularity - 1) / vgprGranularity) * vgprGranularity;

  int64_t sgprGranularity = 8;
  int64_t nextFreeSGPR =
      ((peakSGPRs + sgprGranularity - 1) / sgprGranularity) * sgprGranularity;
  if (llvm::isa<GFX942TargetAttr, GFX950TargetAttr>(targetKind)) {
    nextFreeSGPR = std::min(nextFreeSGPR, int64_t(102));
  }

  if (llvm::isa<GFX942TargetAttr, GFX950TargetAttr>(targetKind)) {
    int64_t accumOffset = std::max(int64_t(4), ((nextFreeVGPR + 3) / 4) * 4);
    accumOffset = std::min(accumOffset, int64_t(256));
    lines.push_back("  .amdhsa_accum_offset " + std::to_string(accumOffset));
    if (nextFreeAGPR > 0) {
      // On CDNA targets AGPRs share the unified file after accum_offset.
      // next_free_vgpr encodes unified usage when AGPRs are present.
      nextFreeVGPR = accumOffset + nextFreeAGPR;
    }
  }

  lines.push_back("  .amdhsa_next_free_vgpr " + std::to_string(nextFreeVGPR));
  lines.push_back("  .amdhsa_next_free_sgpr " + std::to_string(nextFreeSGPR));

  // Always enable all workgroup IDs when any is used.
  // This matches the real LLVM backend and ensures the SGPR layout
  // is predictable (base+0=x, base+1=y, base+2=z) without gaps.
  bool anyWgId = usesWorkgroupIdX || usesWorkgroupIdY || usesWorkgroupIdZ;
  lines.push_back("  .amdhsa_system_sgpr_workgroup_id_x " +
                  std::to_string(anyWgId ? 1 : 0));
  lines.push_back("  .amdhsa_system_sgpr_workgroup_id_y " +
                  std::to_string(anyWgId ? 1 : 0));
  lines.push_back("  .amdhsa_system_sgpr_workgroup_id_z " +
                  std::to_string(anyWgId ? 1 : 0));

  // Derive system_vgpr_workitem_id from workgroup dimensions.
  // When wgZ > 1, hardware provides thread IDs in v0 (x), v1 (y), v2 (z).
  // When wgY > 1 (and wgZ == 1), hardware provides v0 (x), v1 (y).
  int64_t wgY = 1, wgZ = 1;
  auto wgAttr = program.getWorkgroupSizeAttr();
  if (wgAttr) {
    if (wgAttr.size() >= 2)
      if (auto intAttr = dyn_cast<IntegerAttr>(wgAttr[1]))
        wgY = intAttr.getInt();
    if (wgAttr.size() >= 3)
      if (auto intAttr = dyn_cast<IntegerAttr>(wgAttr[2]))
        wgZ = intAttr.getInt();
  }
  int64_t systemVgprWorkitemId = 0;
  if (wgZ > 1) {
    systemVgprWorkitemId = 2;
  } else if (wgY > 1) {
    systemVgprWorkitemId = 1;
  }
  lines.push_back("  .amdhsa_system_vgpr_workitem_id " +
                  std::to_string(systemVgprWorkitemId));

  lines.push_back("  .amdhsa_float_denorm_mode_32 3");
  lines.push_back("  .amdhsa_float_denorm_mode_16_64 3");

  lines.push_back(".end_amdhsa_kernel");

  return lines;
}

llvm::SmallVector<std::string>
MetadataEmitter::emitMetadataYAML(int64_t peakVGPRs, int64_t peakSGPRs,
                                  int64_t peakAGPRs, int64_t ldsSize) {
  llvm::SmallVector<std::string> lines;
  std::string symName = getKernelName(program).str();

  lines.push_back(".amdgpu_metadata");
  lines.push_back("---");
  lines.push_back("amdhsa.version:");
  lines.push_back("  - 1");
  lines.push_back("  - 2");
  lines.push_back("amdhsa.kernels:");
  lines.push_back("  - .name: " + symName);
  lines.push_back("    .symbol: " + symName + ".kd");

  int64_t numArgs = 2;
  if (auto numArgsAttr =
          program->getAttrOfType<IntegerAttr>("num_kernel_args")) {
    numArgs = numArgsAttr.getInt();
  } else {
    int64_t kernargPreload = program.getKernargPreloadLength();
    if (kernargPreload > 0) {
      numArgs = kernargPreload / 2;
    }
  }
  // The runtime packs: buffer_ptrs (8B each) + dynamic_dims (8B each) +
  // stride_args (8B each). Strides are appended by the runtime for each 2D+
  // tensor binding. Use 2x the base size to safely accommodate strides.
  int64_t kernargSize = numArgs * 8 * 2;
  lines.push_back("    .args:");
  for (int64_t i = 0; i < numArgs; ++i) {
    lines.push_back("      - .name:       arg" + std::to_string(i) + "_ptr");
    lines.push_back("        .offset:     " + std::to_string(i * 8));
    lines.push_back("        .size:       8");
    lines.push_back("        .value_kind: global_buffer");
    lines.push_back("        .value_type: 'i8*'");
  }

  lines.push_back("    .kernarg_segment_size: " + std::to_string(kernargSize));
  lines.push_back("    .group_segment_fixed_size: " + std::to_string(ldsSize));
  lines.push_back("    .private_segment_fixed_size: 0");
  lines.push_back("    .kernarg_segment_align: 8");

  int64_t waveSizeVal = target.getDefaultWaveSize();
  lines.push_back("    .wavefront_size: " + std::to_string(waveSizeVal));
  lines.push_back("    .sgpr_count: " + std::to_string(peakSGPRs));
  lines.push_back("    .vgpr_count: " + std::to_string(peakVGPRs));
  if (peakAGPRs > 0)
    lines.push_back("    .agpr_count: " + std::to_string(peakAGPRs));

  auto workgroupSize = program.getWorkgroupSize();
  int64_t maxFlatSize = 256;
  if (workgroupSize.has_value()) {
    maxFlatSize = 1;
    for (Attribute attr : *workgroupSize) {
      if (auto intAttr = dyn_cast<IntegerAttr>(attr)) {
        maxFlatSize *= intAttr.getInt();
      }
    }
  }
  lines.push_back("    .max_flat_workgroup_size: " +
                  std::to_string(maxFlatSize));

  lines.push_back("...");
  lines.push_back(".end_amdgpu_metadata");

  return lines;
}

} // namespace waveasm
